### PR TITLE
lua: timestamp in millis

### PR DESCRIPTION
**Pull request by mindyor**
Commit Message: add Lua function to get timestamp in milliseconds and nanoseconds
Additional Description: Description: adds Lua function to get timestamp in higher resolution. Resolutions are: milliseconds since epoch. Resolution defaults to milliseconds since epoch.
Risk Level: low
Testing: added unit tests for default case, resolution levels, and invalid input
Docs Changes: added documentation for function in `lua_filter.rst`
Release Notes: added release note in `version_history.rst`
Platform Specific Features: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue] #10282
[Optional Deprecated:]

((an update of https://github.com/envoyproxy/envoy/pull/10329))

### CODE REVIEW

**Comment by dio** [comment]
Thanks, @mindyor! Lua API looks good. Could you fix CI?

**Comment by dio** [comment]
@mindyor, do you want me to finish this PR instead? :).

**Comment by dio** [comment]
@mindyor when you have time, please have a look at https://github.com/envoyproxy/envoy/compare/main...dio:lua-timestamp-millis-updated?expand=1. I added two commits (one of them is to "rebase" it with the current main branch "main", the second one is to make it compiles). I think we need to add tests for running `registerGlobal` with some initializers (as use-case tests) in test/extensions/filters/common/lua/lua_test.cc. Hope this helps! Thanks!

**Review by dio — CHANGES_REQUESTED** [review]
Please add tests when calling `registerGlobal` (most likely you need to add it in `test/extensions/filters/common/lua/lua_test.cc`).

File: source/extensions/filters/common/lua/lua.cc [hunk 1]
```diff
@@ -70,10 +70,14 @@ int ThreadLocalState::getGlobalRef(uint64_t slot) {
   return tls.global_slots_[slot];
 }
 
-uint64_t ThreadLocalState::registerGlobal(const std::string& global) {
-  tls_slot_->runOnAllThreads([global](OptRef<LuaThreadLocal> tls) {
+uint64_t ThreadLocalState::registerGlobal(const std::string& global,
+                                          const InitializerList& initializers) {
+  tls_slot_->runOnAllThreads([global, &initializers](OptRef<LuaThreadLocal> tls) {
```
Comment (dio) [inline source/extensions/filters/common/lua/lua.cc:75]: Sorry, my bad, seems like we need to copy the `initializers` vector.

```c++
tls_slot_->runOnAllThreads([global, initializers](OptRef<LuaThreadLocal> tls) {
```

File: source/extensions/filters/common/lua/lua.cc [hunk 2]
```diff
@@ -22,7 +22,7 @@ void Coroutine::start(int function_ref, int num_args, const std::function<void()
 
   state_ = State::Yielded;
   lua_rawgeti(coroutine_state_.get(), LUA_REGISTRYINDEX, function_ref);
-  ASSERT(lua_isfunction(coroutine_state_.get(), -1));
+  ASSERT(lua_isfunction(coroutine_state_.get(), -1) == 1);
```
Comment (dio) [inline source/extensions/filters/common/lua/lua.cc:25]: Can you comment why this needs to be exactly 1? :)

Comment (mindyor) [reply source/extensions/filters/common/lua/lua.cc]: Hmm it may not be. Taking this out

File: source/extensions/filters/common/lua/lua.cc (hunk 1 above)
Comment (mindyor) [reply source/extensions/filters/common/lua/lua.cc]: Thanks! Changed.

**Comment by dio** [comment]
This [`cd870da` (in #14882)](https://github.com/envoyproxy/envoy/pull/14882/commits/cd870dab0b619ce36a9c44d572038f6cea5bc8a6) I think will remedy clang-tidy's complaints.

File: docs/root/configuration/http/http_filters/lua_filter.rst [hunk 3]
```diff
@@ -512,6 +512,17 @@ Encodes the input string as base64. This can be useful for escaping binary data.
...
 
+timestamp()
+^^^^^^^^^^^
+
+.. code-block:: lua
+
+  timestamp = handle:timestamp(format)
+
+High resolution timestamp function. *format* is an optional string parameter to indicate the format of the timestamp.
+*milliseconds_since_epoch* and *nanoseconds_since_epoch* are supported.
```
Comment (dio) [inline docs/root/configuration/http/http_filters/lua_filter.rst:523]: I think we need to document the exported enums here?

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -33,6 +33,7 @@ Minor Behavior Changes
   instead of `//source/common/chromium_url`. The new path canonicalizer is enabled by default. To
   revert to the legacy path canonicalizer, enable the runtime flag
   `envoy.reloadable_features.remove_forked_chromium_url`.
+* lua: added function `timestamp` to provide high resolution timestamps.
```
Comment (dio) [inline docs/root/version_history/current.rst:36]: Could you expand this, so we are aware of the enums.

**Review by dio — CHANGES_REQUESTED** [review]
Looks good. A request to expand the docs a bit more. Thanks!

**Comment by moderation** [comment]
Looks like you need to merge main. Quite a few conflicts

**Review by dio — APPROVED** [review]
Thanks, @mindyor! Looks good!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**